spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnDMLSuite.scala

```diff
@@ -0,0 +1,591 @@
+        checkAnswer(
+          sql(s"SELECT id FROM $table WHERE is_odd = true"),
+          idBeforeReplaceWhere2
+        )
+
+        // IDENTITY VALUES for inserted change records and data should be consistent.
+        checkAnswer(
+          sql(s"SELECT id FROM table_changes('$table', 3, 3) " +
+            "WHERE is_even = true and _change_type = 'insert'"),
+          sql(s"SELECT id FROM $table WHERE is_even = true")
+        )
+
+        // ReplaceWhere source data contains an Identity Column will be blocked.
+        val e = intercept[AnalysisException] {
+          Seq((15, 14), (17, 16)).toDF("id", "value")
+            .withColumn("is_odd", $"value" % 2 =!= 0)
+            .withColumn("is_even", $"value" % 2 === 0)
+            .coalesce(1)
+            .write
+            .format("delta")
+            .mode("overwrite")
+            .option(DeltaOptions.REPLACE_WHERE_OPTION, "is_even = true")
+            .saveAsTable(table)
+        }
+        assert(e.getMessage.contains(
+          "Providing values for GENERATED ALWAYS AS IDENTITY column id is not supported."))
+      }
+    }
+  }
+
+  Seq(true, false).foreach { isPartitioned =>
+    test(s"replaceWhere with CDF enabled [isPartitioned: $isPartitioned]") {
+      testReplaceWhereWithCDF(isPartitioned)
+    }
+  }
+
+  test("CDF for tables with identity column - MERGE") {
+    val src = "identity_test_src"
+    val tgt = "identity_test_tgt"
+    withTable(tgt) {
+      generateTableWithIdentityColumn(tgt)
+      sql(s"ALTER TABLE $tgt SET TBLPROPERTIES (${DeltaConfigs.CHANGE_DATA_FEED.key}=true)")
+
+      withTempView(src) {
+        (1 :: 20 :: 30 :: Nil).toDF("value").createOrReplaceTempView(src)
+        sql(
+          s"""MERGE INTO $tgt target
+             |USING $src src ON target.value = src.value
+             |WHEN MATCHED THEN UPDATE SET target.value = src.value
+             |WHEN NOT MATCHED THEN INSERT (target.value) VALUES (src.value)""".stripMargin)
+      }
+
+      val result = sql(
+        s"""SELECT id, value
+           |FROM $tgt
+           |ORDER BY id""".stripMargin)
+        .as[IdentityColumnTestTableRow]
+        .collect()
+
+      assert(result.length === 8)
+      checkGeneratedIdentityValues(
+        sortedRows = result,
+        start = 0,
+        step = 1,
+        expectedLowerBound = 0,
+        expectedUpperBound =
+          highWaterMark(DeltaLog.forTable(spark, TableIdentifier(tgt)).snapshot, "id"),
+        expectedDistinctCount = 8)
+
+      checkAnswer(sql(
+        s"""SELECT id, value, ${CDCReader.CDC_TYPE_COLUMN_NAME}
+           |FROM table_changes('$tgt', 8)
+           |ORDER BY id, value, _change_type""".stripMargin),
+        Seq(
+          Row(1, 1, CDCReader.CDC_TYPE_UPDATE_POSTIMAGE),
+          Row(1, 1, CDCReader.CDC_TYPE_UPDATE_PREIMAGE),
+          Row(result(6).id, 20, CDCReader.CDC_TYPE_INSERT),
+          Row(result(7).id, 30, CDCReader.CDC_TYPE_INSERT)))
```

Comment:
This is somewhat correct, but not very obvious what it's checking. 

Is it guaranteed that the identity value generated for `20` will be less than the one that is generated for `30`? That seems like an implementation detail that's subject to flakiness. We should be indexing based on what the generated id value was for `20` and `30` instead of assuming.

I suggest making a small helper `getIdForValue(value: long)` which is basically `result.filter(_.value = value).id`. I'd also add a small comment before the checkanswer here that described that it's validating  the ids in the CDCReader matches those in the logical data.